sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamAggregationTransforms.java

```diff
@@ -177,21 +171,30 @@ public AggregationAdaptor(List<AggregateCall> aggregationCalls, RowType sourceRo
           int refIndexKey = call.getArgList().get(0);
           int refIndexValue = call.getArgList().get(1);
 
+          FieldTypeDescriptor keyDescriptor =
+              sourceSchema.getField(refIndexKey).getTypeDescriptor();
           BeamSqlInputRefExpression sourceExpKey = new BeamSqlInputRefExpression(
-                  CalciteUtils.getFieldCalciteType(sourceRowType, refIndexKey), refIndexKey);
+              CalciteUtils.toSqlTypeName(keyDescriptor.getType(), keyDescriptor.getMetadata()),
+              refIndexKey);
+
+          FieldTypeDescriptor valueDescriptor =
+              sourceSchema.getField(refIndexValue).getTypeDescriptor();
           BeamSqlInputRefExpression sourceExpValue = new BeamSqlInputRefExpression(
-                  CalciteUtils.getFieldCalciteType(sourceRowType, refIndexValue), refIndexValue);
+              CalciteUtils.toSqlTypeName(valueDescriptor.getType(), valueDescriptor.getMetadata()),
+                  refIndexValue);
 
           sourceFieldExps.add(KV.of(sourceExpKey, sourceExpValue));
         } else {
           int refIndex = call.getArgList().size() > 0 ? call.getArgList().get(0) : 0;
+          FieldTypeDescriptor typeDescriptor = sourceSchema.getField(refIndex).getTypeDescriptor();
```

Comment:
<!--thread_id:cc_178376529_t; commit:79c95678e593da730ba0472b77304ec1f916245e; resolved:0-->
<!--section:context-quote-->
> **akedin** wrote:
> It's unclear where we're supposed to be using `FieldTypeDescriptor` vs `FieldType`. Can they be combined? So that, for example, all fields in `FieldType` become instances of `FieldTypeDescriptor`. Do we need both?

<!--section:body-->
FieldType is an enum that identifies the type of a row . FieldTypeDescriptor contains extra information needed to resolve the type (e.g. the type of the component element or the schema of the row).

I think it is possible to merge the classes (since Java enums are just classes), but I think it's better to maintain .a clean separation between primitive field types and the recursive spec needed to resolve a type. Also while it's ok to add some extra convenience functionality to an Enum class, making it a fully recursive type seems like an abuse of enums.

For reference, the two classes are _roughly_ equivalent to the SqlTypeName and RelDataType in Calcite. Would it be clearer if I renamed FieldType -> TypeName?